01-getting-started/projects/final/app/src/main/java/com/raywenderlich/podplay/ui/PodcastActivity.kt

```diff
@@ -0,0 +1,312 @@
+    // TODO 2
+    podcastViewModel.viewModelScope.launch(context = Dispatchers.Main) {
+      podcastViewModel.getPodcast(podcastSummaryViewData)
+      hideProgressBar()
+      showDetailsFragment()
+    }
+  }
+
+  private fun scheduleJobs() {
+    val constraints: Constraints = Constraints.Builder().apply {
+      setRequiredNetworkType(NetworkType.CONNECTED)
+      setRequiresCharging(true)
+    }.build()
+
+    // TODO: Chapter 8 - Initially set interval to 5 min to be debugged and fixed with WorkManager
+    val request = PeriodicWorkRequestBuilder<EpisodeUpdateWorker>(
+        1, TimeUnit.HOURS)
+        .setConstraints(constraints)
+        .build()
+
+    WorkManager.getInstance(this).enqueueUniquePeriodicWork(TAG_EPISODE_UPDATE_JOB,
+        ExistingPeriodicWorkPolicy.REPLACE, request)
+  }
+
+  private fun showSubscribedPodcasts() {
+    val podcasts = podcastViewModel.getPodcasts()?.value
+
+    if (podcasts != null) {
+      databinding.toolbar.title = getString(R.string.subscribed_podcasts)
+      podcastListAdapter.setSearchData(podcasts)
+
+      databinding.emptyListViewGroup.visibility = if (podcasts.isEmpty()) View.VISIBLE else View.GONE
+    }
+  }
+
+  private fun performSearch(term: String) {
+    showProgressBar()
+    GlobalScope.launch {
+      val results = searchViewModel.searchPodcasts(term)
+      withContext(Dispatchers.Main) {
+        hideProgressBar()
+        databinding.toolbar.title = term
+        podcastListAdapter.setSearchData(results)
+      }
+    }
+  }
+
+  private fun handleIntent(intent: Intent) {
+    val query = intent.getStringExtra(SearchManager.QUERY) ?: ""
+    performSearch(query)
+
+    // TODO: Chapter 11 - Add checks to prevent unnecessary network calls
+
+    val podcastFeedUrl = intent.getStringExtra(EpisodeUpdateWorker.EXTRA_FEED_URL)
+    if (podcastFeedUrl != null) {
+      podcastViewModel.viewModelScope.launch {
+        val podcastSummaryViewData = podcastViewModel.setActivePodcast(podcastFeedUrl)
+        podcastSummaryViewData?.let { podcastSummaryView -> onShowDetails(podcastSummaryView) }
+      }
+    }
+  }
+
+
+  private fun setupToolbar() {
+    setSupportActionBar(databinding.toolbar)
+  }
+
+  private fun setupViewModels() {
+    val service = ItunesService.instance
+    searchViewModel.iTunesRepo = ItunesRepo(service)
+    val rssService = RssFeedService.instance
+    podcastViewModel.podcastRepo = PodcastRepo(rssService, podcastViewModel.podcastDao)
+  }
+
+  private fun setupPodcastListView() {
+    podcastViewModel.getPodcasts()?.observe(this, {
+      if (it != null) {
+        showSubscribedPodcasts()
+      }
+    })
+  }
+
+  private fun addBackStackListener() {
+    supportFragmentManager.addOnBackStackChangedListener {
+      if (supportFragmentManager.backStackEntryCount == 0) {
+        databinding.podcastRecyclerView.visibility = View.VISIBLE
+      }
+    }
+  }
+
+  private fun updateControls() {
+    databinding.podcastRecyclerView.setHasFixedSize(true)
```

Comment:
Just wondering, I didn't research it - do you know why Kotlin extensions do not work on recyclerView when doing a setup?